dashboard/app/helpers/users_helper.rb

```diff
@@ -168,68 +168,39 @@ def user_summary(user)
       sl.level_ids.each do |level_id|
         # if we have a contained level or BubbleChoice level, use that to represent progress
         level = Level.cache_find(level_id)
-        sublevel_id = level.is_a?(BubbleChoice) ? level.best_result_sublevel(user)&.id : nil
+
         if level.is_a?(BubbleChoice) # we have a parent level
+          sum_time_spent = 0
+
           # get progress for sublevels to save in levels hash
           level.sublevels.each do |sublevel|
             ul = user_levels_by_level.try(:[], sublevel.id)
-            completion_status = activity_css_class(ul)
-            # a UL is submitted if the state is submitted UNLESS it is a peer reviewable level that has been reviewed
-            submitted = !!ul.try(:submitted) &&
-              !(ul.level.try(:peer_reviewable?) && [ActivityConstants::REVIEW_REJECTED_RESULT, ActivityConstants::REVIEW_ACCEPTED_RESULT].include?(ul.best_result))
-            readonly_answers = !!ul.try(:readonly_answers)
-            locked = ul.try(:show_as_locked?, sl.lesson) || sl.lesson.lockable? && !ul
-            if completion_status == LEVEL_STATUS.not_tried
-              # for now, we don't allow authorized teachers to be "locked"
-              if locked && !user.authorized_teacher?
-                progress[level_id] = {
-                  status: LEVEL_STATUS.locked
-                }
-              end
-              next
-            end
-            progress[sublevel.id] = {
-              status: completion_status,
-              result: ul.try(:best_result) || 0,
-              submitted: submitted ? true : nil,
-              readonly_answers: readonly_answers ? true : nil,
-              paired: (paired_user_levels.include? ul.try(:id)) ? true : nil,
-              locked: locked ? true : nil,
-              last_progress_at: include_timestamp ? ul&.updated_at&.to_i : nil,
-              time_spent: ul&.time_spent&.to_i
-            }.compact
+            sublevel_progress = get_level_progress(ul, sl, paired_user_levels, include_timestamp)
+            progress[sublevel.id] = sublevel_progress.compact
+            sum_time_spent += sublevel_progress[:time_spent] || 0
           end
+
+          best_sublevel_id = level.best_result_sublevel(user)&.id
```

Comment:
if you do move the `BubbleChoice` block out into a helper, i think it would be worth moving this line to the top and returning early if it returns nil -- more for readability than for performance